1666: Immutable state for cards and extensions

steffenkleinle opened the pull request:
### Short description

<!-- Describe this PR in one or two sentences. -->
As discussed in #1666, mutating react state is bad practice and commonly leads to bugs as react is not aware of any changes.
Furthermore, mutating objects in place is generally not advisable as it introduces side effects and is also a source of bugs and makes the code less understandable.

This PR aims to address this issues in the case of cards and extensions by splitting the pure state (to be used in useStates/hooks) and helper methods around it to construct/mutate/check this state.

### Proposed changes

<!-- Describe this PR in more detail. -->

- Use simple objects for extensions
- Use correct react components for extension forms
- Use simple objects for cards (blueprints)
- Refactoring helper functions to reflect these changes

### Resulting Improvements

- No more `notifyUpdate` necessary
- Simple and easy adjustment of state (typesafe)
- Easy accessing of extension values

### Side effects

<!-- List all related components that have not been explicitly changed but may be affected by this PR -->
Hopefully none?

### Testing
1. Single card creation for bavaria, nuernberg and koblenz (/erstellen)
2. Create cards from link with queryParams find examples here: https://github.com/digitalfabrik/entitlementcard/pull/1089 Please also check correct addressData on pdf.
3. Generate cards via csv for nuernberg and bavaria. Find bulk import files here: (administration/resources/cards)
4. Import csv from freinet (bavaria)
5. Test csv export for nuernberg
6. Please also activate these cards to check if everything works with the qrCodes and deeplinks

<!-- List all things that should be tested while reviewing this PR. -->

### Resolved issues

<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #1666.

steffenkleinle commented on administration/src/bp-modules/cards/ImportCardsController.tsx:
```diff
@@ -59,23 +59,15 @@ const InnerImportCardsController = ({ region }: { region: Region }): ReactElemen
-      headers.forEach(header => {
-        const idx = csvHeader.indexOf(header)
-        if (idx === -1) {
-          // column is missing in csv
-          return
-        }
-        cardBlueprint.setValue(header, line[idx])
-      })
-      return cardBlueprint
+      return initializeCardFromCSV(projectConfig.card, line, csvHeader, region)
```
This is still really confusing, why are haders and csvHeader mixed here? Which one do I have to use?

f1sh1918 replied:
we have headers defined in the config  (headers) and the headers that were read from the csv (csvHeaders).

steffenkleinle replied:
I will create a refactoring issue as this is really confusing...

steffenkleinle replied:
https://github.com/digitalfabrik/entitlementcard/issues/1700

steffenkleinle commented on administration/src/bp-modules/self-service/hooks/useCardGeneratorSelfService.ts:
```diff
@@ -27,26 +25,25 @@ type UseCardGeneratorSelfServiceReturn = {
   generateCards: () => Promise<void>
   downloadPdf: (code: CreateCardsResult, fileName: string) => Promise<void>
 }
-const initializeCardBlueprintForProject = (projectConfig: ProjectConfig): SelfServiceCard => {
-  const headers = getHeaders(projectConfig)
-  const selfServiceCard = new SelfServiceCard('', projectConfig.card)
-  headers.forEach(header => {
-    selfServiceCard.setValue(header, '')
-  })
-  return selfServiceCard
-}
```
Why was this necessary? It is just setting empty values anyway?

f1sh1918 replied:
because of the `isValid()` and `isFullnameValid()` fkts. Currently they will be executed directly because we have no untouched state

steffenkleinle commented on administration/src/cards/CsvFactory.ts:
```diff
@@ -37,8 +36,6 @@ export function generateCsv(
 
 export const getCSVFilename = (cardBlueprints: CardBlueprint[]): string => {
   const filename =
-    cardBlueprints.length === 1
-      ? findExtension(cardBlueprints[0].extensions, NuernbergPassIdExtension)?.state?.passId
-      : 'Pass-ID[0]mass'
```
This only works for nuernberg, does it? What is the file name for other projects?

f1sh1918 replied:
we don't have csv export for other projects, so no filename needed or what do you mean?

steffenkleinle replied:
Ahh thanks for the explanation :+1:

f1sh1918 replied:
but actually this fallback is ugly i would use sth like `SozialpassMassExport`

steffenkleinle commented on administration/src/cards/SelfServiceCard.ts:
```diff
@@ -1,10 +0,0 @@
-import { CardBlueprint } from './CardBlueprint'
-
-class SelfServiceCard extends CardBlueprint {
-  // Override isValid() to remove expiration check
-  isValid(): boolean {
-    return this.isFullNameValid() && this.extensions.every(ext => ext.isValid())
-  }
```
Why?

f1sh1918 replied:
because it will be added in the backend. We don't have information in the frontend for that

steffenkleinle commented on administration/src/cards/CardBlueprint.ts:
```diff
@@ -5,144 +5,205 @@ import { CardExtensions, CardInfo } from '../generated/card_pb'
+  region: Region | undefined = undefined,
+  { id, fullName, expirationDate, extensions }: Partial<CardBlueprint> = {}
+): CardBlueprint => {
+  // TODO Why only if region is defined?
+  // const defaultExpirationDate = region ? PlainDate.fromLocalDate(new Date()).add(cardConfig.defaultValidity).toDaysSinceEpoch() : null
```
Why are we setting a expiration date only if a region is passed?

f1sh1918 replied:
tbh i have no idea. i was also really confused about this initialization stuff.
We have to pass this region in general because it will not be provided by form input nor csv file. And we get it from login information

steffenkleinle replied:
Okay, I'll change this to always initialize it.

steffenkleinle commented on administration/src/cards/__tests__/SelfServiceCard.test.ts:
```diff
@@ -1,30 +1,45 @@
+    expect(generateCardInfo(card).toJson()).toEqual({
       fullName: 'Karla Koblenz',
-      extensions: {},
+      // TODO why should this be empty/null?
```
Why should a newly initialized card have no values set yet?

steffenkleinle commented on administration/src/cards/extensions/AddressFieldExtensions.ts:
```diff
@@ -1,45 +1,47 @@
+  Component: () => null,
+  getInitialState: () => ({ [name]: '' } as AddressFieldExtensionState<T>),
+  causesInfiniteLifetime: () => false,
+  // TODO Why no protobuf data?
```
Is it correct that address fields (and other extensions) are not included in protobuf data?

f1sh1918 replied:
it depends on the extension if its included
```js
  setProtobufData(message: PartialMessage<CardExtensions>): void {
    message.extensionBirthday = {
      birthday: this.state?.birthday,
    }
  }
```

steffenkleinle replied:
Yes, sure. But why is the address not included for example?

f1sh1918 replied:
because its not needed in the qr code. this is just for the pdf letterhead and will not be transferred to the app.

steffenkleinle commented on administration/src/cards/extensions/BirthdayExtension.tsx:
```diff
@@ -1,99 +1,73 @@
-  public readonly name = BirthdayExtension.name
+const BirthdayForm = ({ value, setValue, isValid }: ExtensionComponentProps<BirthdayExtensionState>): ReactElement => {
+  const { viewportSmall } = useWindowDimensions()
+  const inputColor = value === getInitialState() ? Colors.GRAY1 : Colors.BLACK
```
Shouldn't we rather set `null` as initial state and render `01.01.1900` as placeholder?

f1sh1918 replied:
we still have design issues for that

f1sh1918 replied:
the problem is that then the current date is used if you open the datepicker. we can do that but when i remember right the more annoying thing was to the scroll hell if not setting a default value to sth closer to most common birthdays

steffenkleinle commented on administration/src/cards/extensions/EMailNotificationExtension.ts:
```diff
@@ -1,24 +1,19 @@
+const EMailNotificationExtension: Extension<EmailNotificationExtensionState> = {
+  name: EMAIL_NOTIFICATION_EXTENSION_NAME,
+  Component: () => null,
+  // TODO Why initial null?
+  getInitialState: () => ({ [EMAIL_NOTIFICATION_EXTENSION_NAME]: '' }),
+  causesInfiniteLifetime: () => false,
+  // TODO Why no protobuf data?
```
Why is this initially set to null?

f1sh1918 replied:
if it will not be passed as a queryParam or from the applications view it will be null since we have no information about email in some cases when cards will be created directly without application (nuernberg project)

steffenkleinle commented on administration/src/cards/CardBlueprint.ts:
```diff
@@ -5,144 +5,205 @@ import { CardExtensions, CardInfo } from '../generated/card_pb'
-        this.extensions[extensionIdx].fromString(value)
-    }
-  }
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
I was not able to construct the correct type, feel free to try :/

f1sh1918 replied:
hm i removed the comment and don't get any error

ztefanie replied:
i think this is the correct type:
`Array<{
  extension: Extension<InferExtensionStateType<typeof Extensions[number]>>;
  state: InferExtensionStateType<typeof Extensions[number]>;
}>`

steffenkleinle replied:
Thank you @ztefanie, nice! :tada:

steffenkleinle commented on administration/src/cards/CardBlueprint.ts:
```diff
@@ -5,144 +5,205 @@ import { CardExtensions, CardInfo } from '../generated/card_pb'
+  const startDay = cardBlueprint.extensions.startDay
+  const startDayDate = startDay !== undefined ? PlainDate.fromDaysSinceEpoch(startDay) : null
+
+  // TODO Why should this not be checked for self service? Why is this allowed to be nullable for csv/self service?
+  // TODO Adjust isValid
```
See other question, why is this allowed to be nullable in some cases? Why shouldn't this be checked for self service cards?

f1sh1918 replied:
hm i think the name shouldn't be nullable and the extension depend on which extension i think.
For selfServiceCards we have no expiration day in frontend

steffenkleinle commented:
We are mixing the names `Card` and `CardBlueprint` quite a lot. What is the meaning of the `blueprint` in this case? Can't we just use `card` consistently?

f1sh1918 commented:
> We are mixing the names `Card` and `CardBlueprint` quite a lot. What is the meaning of the `blueprint` in this case? Can't we just use `card` consistently?

for me card would be fine

michael-markl commented on administration/src/util/helper.ts:
```diff
@@ -1,2 +1,11 @@
 export const isDevMode = (): boolean => window.location.hostname === 'localhost'
 export const isStagingMode = (): boolean => !!window.location.hostname.match(/staging./)
+
+export const updateArrayItem = <T>(array: T[], updatedItem: T, index: number): T[] => [
+  ...(index > 0 ? array.slice(0, index) : []),
+  updatedItem,
+  ...(index < array.length - 1 ? array.slice(index + 1) : []),
+]
```
I think it would make sense here to throw an Error if index is not in range.

steffenkleinle commented on administration/src/cards/CardBlueprint.ts:
```diff
@@ -5,144 +5,203 @@ import { CardExtensions, CardInfo } from '../generated/card_pb'
-  expirationDate: string | null
-  extensions: (E extends Extension<infer T, unknown> ? JSONExtension<T> : never)[]
+  // TODO null necessary? csv?
+  expirationDate: number | null
```
Yes, and apparently apart from this it is necessary for self service cards. Not sure if it wouldn't make more sense to refactor it as an extension then too, though.

steffenkleinle commented:
@michael-markl sorry for rebasing, I did not realize you are currently reviewing.

f1sh1918 commented:
Could test that much @steffenkleinle when i click on create card i get this error

<img width="1439" alt="image" src="https://github.com/user-attachments/assets/ca8163ba-9c95-42c1-862d-1b6a15404b6c">

f1sh1918 commented:
i added the most important testing scenarios @steffenkleinle 
If i forgot sth @ztefanie please feel free to add

michael-markl commented on administration/src/cards/CardBlueprint.ts:
```diff
@@ -5,144 +5,195 @@ import { CardExtensions, CardInfo } from '../generated/card_pb'
+type CardBlueprintKey = keyof Omit<Omit<CardBlueprint, 'extensions'>, 'id'>
+type KeyType = CardBlueprintKey | ExtensionKey | null
+
+const createRandomId = () => Math.floor(Math.random() * 1000000)
```
It's quite likely that there will be collisions. if we generate IDs with only 1,000,000 buckets. With 1000 cards, the probability that there is a collision is roughly 40% (see [birthday problem](https://en.wikipedia.org/wiki/Birthday_problem))

What are these ids used for ? Is it possible to just use global counter?

steffenkleinle replied:
Actually not sure, I just kept this from the previous implementation without any logical changes. I was wondering the same though... Not sure what they are used for.

michael-markl replied:
Alright, maybe we should create an issue then. Seems quite suspicious.

f1sh1918 commented:
merged recent main to get rid of db migration problems

f1sh1918 reviewed (changes requested):
One thing i forgot in the testing instructions @steffenkleinle 
Nuernberg also exports csv data to send it to a printer service from datev.
The export leads to wrong dates
Second table entry shows the correct output

<img width="1047" alt="Bildschirmfoto 2024-10-22 um 10 59 07" src="https://github.com/user-attachments/assets/f13d61fb-2949-431e-af86-748265fa8472">

Another issue i faced that i can't create cards for koblenz.
I get "Ein unbekannter Fehler ist aufgetreten" even i type i correct data.
And i face some weird issued in the backend log for value encoding

```sh
SQL: INSERT INTO cards ("activationSecretHash", "cardInfoHash", "codeType", "entitlementId", "expirationDay", "issueDate", "regionId", revoked, "startDay") VALUES ($2a$11$IjK19SdDJNL3tevl9MkNZ.NDSJFAeS0kD7e3XlHUzmpzAooX0v1eu, �$���|v���gR�9�� nU�Ɩ���7�, 1, 27, 20089, '2024-10-22T09:14:13.499388', 95, FALSE, 19723)
SQL: INSERT INTO cards ("activationSecretHash", "cardInfoHash", "codeType", "entitlementId", "expirationDay", "issueDate", "regionId", revoked, "startDay") VALUES (NULL, ���M��h0>���̵>�C�}4D6�Pp1�Ǥ\, 0, 27, 20089, '2024-10-22T09:14:13.529543', 95, FALSE, 19723)
```
I was facing this also when i switched from an array to a single card blueprint object when i tried to implement this first and then switched back to that array implementation. But no idea why this happens.
I think sth doesn't work correctly while creating hash

steffenkleinle commented:
> One thing i forgot in the testing instructions @steffenkleinle Nuernberg also exports csv data to send it to a printer service from datev. The export leads to wrong dates Second table entry shows the correct output

Should be fixed.

> Another issue i faced that i can't create cards for koblenz. I get "Ein unbekannter Fehler ist aufgetreten" even i type i correct data. And i face some weird issued in the backend log for value encoding

In the self service portal? I can't create/import cards with any user as `add/cards` is not available for me. For the self service portal, everything works for me.

> 
> ```shell
> SQL: INSERT INTO cards ("activationSecretHash", "cardInfoHash", "codeType", "entitlementId", "expirationDay", "issueDate", "regionId", revoked, "startDay") VALUES ($2a$11$IjK19SdDJNL3tevl9MkNZ.NDSJFAeS0kD7e3XlHUzmpzAooX0v1eu, �$����|v���gR��9����nU�Ɩ���7�, 1, 27, 20089, '2024-10-22T09:14:13.499388', 95, FALSE, 19723)
> SQL: INSERT INTO cards ("activationSecretHash", "cardInfoHash", "codeType", "entitlementId", "expirationDay", "issueDate", "regionId", revoked, "startDay") VALUES (NULL, ���M��h0>����̵>�C�}4D6�Pp1�Ǥ�\, 0, 27, 20089, '2024-10-22T09:14:13.529543', 95, FALSE, 19723)
> ```
>
> I was facing this also when i switched from an array to a single card blueprint object when i tried to implement this first and then switched back to that array implementation. But no idea why this happens. I think sth doesn't work correctly while creating hash

What exactly is weird/wrong about that? I just tested with main and I don't see a difference here...

f1sh1918 commented:
@steffenkleinle 
Looks like bulk import is now broken.
If i make a csv import for nuernberg or bavaria and try to export csv or create cards i get "Es ist ein unbekannter Fehler aufgetreten"

f1sh1918 reviewed (approved):
Tested all use cases. works great. 👍

ztefanie commented on administration/src/util/PlainDate.ts:
```diff
@@ -56,11 +58,26 @@ class PlainDate {
   }
 
+  static safeFromCustomFormat(value: string | null, format: string = 'dd.MM.yyyy'): PlainDate | null {
+    if (value === null) {
+      return null
+    }
+    try {
+      return PlainDate.fromCustomFormat(value, format)
+    } catch {
+      return null
+    }
```
```suggestion
return value && PlainDate.fromCustomFormat(value, format).catch(() => null);
```

steffenkleinle replied:
Sadly this does not work as
a) `value &&` might return empty strings `""`
b) `.catch()` does not exist on non-promise values

ztefanie commented on administration/src/util/helper.ts:
```diff
@@ -1,2 +1,16 @@
+    throw new RangeError(`Index ${index} is out of range`)
+  }
+  return [
+    ...(index > 0 ? array.slice(0, index) : []),
+    updatedItem,
+    ...(index < array.length - 1 ? array.slice(index + 1) : []),
```
```suggestion
    ...array.slice(0, index),
    updatedItem,
    ...array.slice(index + 1),
```

ztefanie reviewed (commented):
I did review this, but to be honest I did not really fully understand what you did here :-(
But everything looks fine to me and like a good improvement.
I added some small adjustment suggestions.

Tested works as expected

steffenkleinle commented:
> I did review this, but to be honest I did not really fully understand what you did here :-( But everything looks fine to me and like a good improvement. I added some small adjustment suggestions.
> 
> Tested works as expected

Please have a look at the issue, I think its mostly explained there. But basically, updated state is only recognized by react if the value does not shallow equal (i.e. it is the same object reference) as before. This prevents things like `cardBlueprint.expirationDate = newExpirationDate` as we did before. Therefore I replaced the usages of the classes `CardBlueprint` and `Extension` with corresponding plain objects.
If you have any other questions, feel free to ask and I can explain more in detail or in a call.

michael-markl commented:
I get a validation error on this MailNotification column (Bayern). However, clicking on "QR Codes drucken" still works.

Also, it's weird that the last line does not give a validation error as it's a Goldkarte and according to the table has a expiration date, although Goldkarten do not have expiration dates. Also 01.01.0025 looks like a very invalid date :D

![grafik](https://github.com/user-attachments/assets/5e226084-af91-4dd8-b55f-ad1e40c68a9d)

f1sh1918 commented:
> I get a validation error on this MailNotification column (Bayern). However, clicking on "QR Codes drucken" still works.
> 
> Also, it's weird that the last line does not give a validation error as it's a Goldkarte and according to the table has a expiration date, although Goldkarten do not have expiration dates. Also 01.01.0025 looks like a very invalid date :D
> 
> ![grafik](https://github.com/user-attachments/assets/5e226084-af91-4dd8-b55f-ad1e40c68a9d)
> 

Thx for testing. These issues existed before. I think we should remove the Mail notification field for csv. I will create an issue for that even it could also be used as a feature in the future so send mails with deeplinks. That should be discussed.

I think since it's a goldcard it will be ignored. Is not really good ux though. You can create a ticket
